File: csharp/src/code-smells/change-preventers/ShotgunSurgery.cs
namespace RefactoringAvanzado.CodeSmells.ChangePreventers;

public record LineItem(string Name, double Price, int Qty);

public class PriceCalculator
{
    public double TotalWithTax(LineItem[] items)
    {
        var subtotal = items.Sum(i => i.Price * i.Qty);
        var tax = subtotal * 0.21;
        return subtotal + tax;
    }
}

public class InvoiceService
{
    public double CreateTotal(LineItem[] items)
    {
        var baseAmount = items.Sum(i => i.Price * i.Qty);
        var vat = baseAmount * 0.21;
        return baseAmount + vat;
    }
}

public class SalesReport
{
    public string Summarize(LineItem[] items)
    {
        var sum = items.Sum(i => i.Price * i.Qty);
        var tax = sum * 0.21;
        var total = sum + tax;
        return $"total={total:F2}";
    }
}

public class LoyaltyPoints
{
    public int Points(LineItem[] items)
    {
        var baseAmount = items.Sum(i => i.Price * i.Qty);
        var withTax = baseAmount + baseAmount * 0.21;
        return (int)Math.Floor(withTax / 10);
    }
}

public static class ShotgunDemo
{
    public static (double, double) DemoShotgun(LineItem[] items)
    {
        var calc = new PriceCalculator().TotalWithTax(items);
        var inv = new InvoiceService().CreateTotal(items);
        return (calc, inv);
    }
}

File: typescript/src/code-smells/change-preventers/shotgun-surgery.ts
export type LineItem = { name: string; price: number; qty: number }

export class PriceCalculator {
  totalWithTax(items: LineItem[]): number {
    const subtotal = items.reduce((s, i) => s + i.price * i.qty, 0)
    const tax = subtotal * 0.21
    return subtotal + tax
  }
}

export class InvoiceService {
  createTotal(items: LineItem[]): number {
    const base = items.reduce((s, i) => s + i.price * i.qty, 0)
    const vat = base * 0.21
    return base + vat
  }
}

export class SalesReport {
  summarize(items: LineItem[]): string {
    const sum = items.reduce((s, i) => s + i.price * i.qty, 0)
    const tax = sum * 0.21
    const total = sum + tax
    return `total=${total.toFixed(2)}`
  }
}

export class LoyaltyPoints {
  points(items: LineItem[]): number {
    const base = items.reduce((s, i) => s + i.price * i.qty, 0)
    const withTax = base + base * 0.21
    return Math.floor(withTax / 10)
  }
}

export function demoShotgun(items: LineItem[]): [number, number] {
  const calc = new PriceCalculator().totalWithTax(items)
  const inv = new InvoiceService().createTotal(items)
  return [calc, inv]
}
